cluster-autoscaler/core/equivalence_groups.go

```diff
@@ -87,3 +87,22 @@ func groupPodsBySchedulingProperties(pods []*apiv1.Pod) map[equivalenceGroupId][
 
 	return podEquivalenceGroups
 }
+
+func podSpecSemanticallyEqual(p1 apiv1.PodSpec, p2 apiv1.PodSpec) bool {
+	p1Spec := sanitizeVolumesAndMounts(p1)
+	p2Spec := sanitizeVolumesAndMounts(p2)
+	return apiequality.Semantic.DeepEqual(p1Spec, p2Spec)
+}
+
+// sanitizeVolumesAndMounts removes volumes and volume mounts from the pod spec
+// Due to the generated suffixes, a strict DeepEquals check will fail and generate
+// an equivalence group per pod which is undesirable.
+// This is also applicable to a pod without volumes scenario due to the volume
+// projection of the default service account onto the kube-api-access volume
+func sanitizeVolumesAndMounts(podSpec apiv1.PodSpec) apiv1.PodSpec {
+	podSpec.Volumes = nil
```

Comment:
Caching result of scheduler Filters only makes sense if we know the pods are identical as far as scheduling is concerned. As far as I understand projected volumes don't affect scheduling, so it makes sense to ignore them in comparison function. 
However, PVs impact scheduling. If pod is using PVC that already has a bound PV the zone where a pod needs to run is determined by where the PV is. IIRC this would generally be the case when using Immediate volumeBindingMode. Even when using WaitForFirstConsumer we should make sure that allowedTopologies match. 

So I think this implementation is overly broad. We should narrow it down to make sure the pods in equivalence group really are identical as far as scheduling is concerned. As much as I dislike the idea, the most obvious way to do this seem to be by selectively filtering out volumes that don't affect scheduling.

Reply:
Agreed on the broad comment. We can filter out the projected volumes by name then null out the volume mounts in the containers.

One question though - doesn't this mean the performance with SatetfulSets is subpar? Every pod will get its own claim name so a large scale PVC scenario, building this equivalence group can take forever. I understand its necessary because the scheduler needs to consider those PVC constraints which CA has no visibility over. But, is there something better we can do? Maybe short-circuit building this method somehow?

Reply:
Replying to the question:
 * Conceptually in scale-up we run Filters in two places: [initial filtering](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/scale_up.go#L294) (check if pods fit on an empty node in a given nodegroup) and later in [binpacking](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/estimator/binpacking_estimator.go#L83). In each step we run PreFilters once per pod, in initial filtering we also run Filters once, while in binpacking we run it O(node_delta).
 * Equivalence groups are just and optimization for the initial filtering. Since we're doing check for each pod independently we opportunistically cache the results and re-use them for identical pods. This idea doesn't really work for StatefulSets, since every pod in StatefulSet has a different PVC and so may need to run in a different zone (and since binpacking is done on a zonal NodeGroup that means that the result of initial filtering should actually be different between pods in SS).
   * Note that this optimization reduces time spent running PreFilters/Filters in scale-up by at most 50% (as explained above equivalence groups don't feature in actual binpacking and we run binpacking simulation per-NodeGroup, not per-EquivalenceGroup). We also run PreFilters/Filters [earlier in CA loop](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/filter_out_schedulable.go#L114) as well as do a bunch of API calls to provider and other time consuming operations. So, while significant, this optimization shouldn't make or break CA performance. If it does can you share the logs and function_duration_seconds metric for an example run? 

The problem with StatefulSets goes deeper than this though. CA does entire binpacking using a NodeGroup in one zone and, [if all the pods in scale-up can run in any zone](https://github.com/kubernetes/autoscaler/blob/821d6771b0222d3d84a76717065db4f53e3ea026/cluster-autoscaler/core/scale_up.go#L660) it splits this scale-up between zones. If you have a StatefulSet that uses Immediate volumeBindingMode and gets spread over 3 zones, the individual pods in StatefulSet are no longer 'zone-agnostic' and we can't just calculate total number of nodes needed and arbitrarily split it between zones. Instead a zonal binpacking will only accomodate pods in one zone (ie. 1/3 of all pods) and the rest of pods will only be handled by next run of scale-up logic. In other words for StatefulSet you would generally need 3x more autoscaler loops (swap 3 with the number of zones StatefulSet can schedule into). 

So StatefulSets are expected to be much slower, but I don't think this particular piece of logic should be the primary reason for it. However, I suspect the first version of your change may produce much shorter latency for StatefulSets, particularly if you test by just creating a large SS in one go. 
I think what happens is your logic combines with bug #4129. By placing all the pods in StatefulSet in the same equivalence group you're essentially bypassing the initial filtering phase (it's enough for one pod to fit in a given zone for CA to decide they all fit). #4129 is caused by the fact that binpacking logic never rechecks initial filtering logic and just blindly assumes that anything that passed initial filtering would fit on a new node. This means that any StatefulSet pod that can't run in the zone used in binpacking is translated into an extra node in scale-up decision (even if multiple pods would fit on a node, each will get a separate node!). CA will overshoot scale-up by a large factor on every loop, which negates the need to run extra loops described above and obviously helps the latency a lot. In a single, large scale-up test the end result may look reasonable - CA starts scaling-up before all the pods are created, so the fact that CA adds way too many nodes initially is hidden by the fact that more pods are created later on and schedule on those node.

To sum up - the fact that pods from StatefulSet gets into different equivalence groups is I think correct, because they're _not_ equivalent for scheduling purposes. Poor StatefulSet comes from how CA handle multi-zonal cluster and would probably require doing multi-zonal binpacking (ie. completely rewriting scale-up...) to fix.

Reply:
Also - a general recommendation for any PV is to never use Immediate volumeBindingMode with CA and instead use WaitForFirstConsumer. If I understand correctly this way StatefulSets are pod are zone agnostic and can all be handled in a single scale-up. They will still end up in separate equivalence groups, but that should not have as big of an impact by itself (it just requires one extra [CheckPredicates call](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/scale_up.go#L294) for each pod, but shouldn't impact binpacking).

Reply:
Thanks for the lengthy response on the stateful set case.

The test case (and what the PR is targeting) is a simple nginx deployment. It made me think about the SS case but its tangential.

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  selector:
    matchLabels:
      run: nginx
  replicas: 1
  template:
    metadata:
      labels:
        run: nginx
    spec:
      containers:
      - name: nginx
        image: docker.io/library/nginx:1.13.12-alpine
        ports:
        - containerPort: 80
        resources:
          limits:
            cpu: 500m
          requests:
            cpu: 200m
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    run: nginx
spec:
  ports:
  - port: 80
  selector:
    run: nginx
```
 - CA changes:
 ```git
diff --git i/cluster-autoscaler/core/equivalence_groups.go w/cluster-autoscaler/core/equivalence_groups.go
--- i/cluster-autoscaler/core/equivalence_groups.go
+++ w/cluster-autoscaler/core/equivalence_groups.go
@@ -17,7 +17,10 @@ limitations under the License.
 package core

 import (
+	"k8s.io/autoscaler/cluster-autoscaler/metrics"
+	"k8s.io/klog/v2"
 	"reflect"
+	"time"

 	apiv1 "k8s.io/api/core/v1"
 	apiequality "k8s.io/apimachinery/pkg/api/equality"
@@ -34,6 +37,8 @@ type podEquivalenceGroup struct {

 // buildPodEquivalenceGroups prepares pod groups with equivalent scheduling properties.
 func buildPodEquivalenceGroups(pods []*apiv1.Pod) []*podEquivalenceGroup {
+	klog.V(1).Info("At the start of buildPodEquivalence Group")
+	buildPodEquivalenceGroupStart := time.Now()
 	podEquivalenceGroups := []*podEquivalenceGroup{}
 	for _, pods := range groupPodsBySchedulingProperties(pods) {
 		podEquivalenceGroups = append(podEquivalenceGroups, &podEquivalenceGroup{
@@ -42,6 +47,8 @@ func buildPodEquivalenceGroups(pods []*apiv1.Pod) []*podEquivalenceGroup {
 			schedulable:      false,
 		})
 	}
+     klog.V(1).Infof("Done computing equivalence groups with length: %s", len(podEquivalenceGroups))
+	metrics.UpdateDurationFromStart("buildPodEquivalenceGroups", buildPodEquivalenceGroupStart)
 	return podEquivalenceGroups
 }
diff --git i/cluster-autoscaler/metrics/metrics.go w/cluster-autoscaler/metrics/metrics.go
--- i/cluster-autoscaler/metrics/metrics.go
+++ w/cluster-autoscaler/metrics/metrics.go
@@ -277,9 +277,9 @@ func UpdateDurationFromStart(label FunctionLabel, start time.Time) {
 func UpdateDuration(label FunctionLabel, duration time.Duration) {
 	// TODO(maciekpytel): remove second condition if we manage to get
 	// asynchronous node drain
-	if duration > LogLongDurationThreshold && label != ScaleDown {
-		klog.V(4).Infof("Function %s took %v to complete", label, duration)
-	}
+	//if duration > LogLongDurationThreshold && label != ScaleDown {
+		klog.V(1).Infof("Function %s took %v to complete", label, duration)
+	//}
 	functionDuration.WithLabelValues(string(label)).Observe(duration.Seconds())
 	functionDurationSummary.WithLabelValues(string(label)).Observe(duration.Seconds())
 }
```

- Start CA with 40k pods **already** pending, no cpu-limits, scan-interval of 60s. I understand that an incremental run where pods are coming up gradually vs a one-shot large pending pods will perform differently but I think the bottlenecks here apply to both cases. 
- There's two issues here I've hit:
  - The initial filtering `filterOutSchedulablePodListProcessor` takes a long time to compute which is expected since it would run the filters per pod. I haven't gotten to digging into it what we can optimize there yet (Can we build pod equivalency and possibly exit early instead of running 40k times?). You can wait for it to complete or just return `unschedulablePods` from it since that would be the end result anyways if you are starting CA at this point in time.
  - `buildPodEquivalenceGroups` takes a long time as well prior to the changes in this PR.
 
- **First Run without the changes:**
```
I1109 17:44:48.843677       1 metrics.go:281] Function updateClusterState took 1.63602ms to complete
I1109 17:44:48.889530       1 filter_out_schedulable.go:64] Filtering out schedulables
I1109 18:34:55.929658       1 metrics.go:281] Function filterOutSchedulable took 50m7.040078821s to complete
I1109 18:34:55.929766       1 filter_out_schedulable.go:81] No schedulable pods
I1109 18:34:55.959080       1 klogx.go:86] Pod default/nginx-7485d99856-p9ww8 is unschedulable
I1109 18:34:55.959143       1 klogx.go:86] Pod default/nginx-7485d99856-5b577 is unschedulable
I1109 18:34:55.959154       1 klogx.go:86] Pod default/nginx-7485d99856-vx8h5 is unschedulable
I1109 18:34:55.959162       1 klogx.go:86] Pod default/nginx-7485d99856-m7ndx is unschedulable
I1109 18:34:55.959169       1 klogx.go:86] Pod default/nginx-7485d99856-xwn56 is unschedulable
I1109 18:34:55.959177       1 klogx.go:86] Pod default/nginx-7485d99856-qmqcf is unschedulable
I1109 18:34:55.959184       1 klogx.go:86] Pod default/nginx-7485d99856-pwb9w is unschedulable
I1109 18:34:55.959192       1 klogx.go:86] Pod default/nginx-7485d99856-pwjw9 is unschedulable
I1109 18:34:55.959199       1 klogx.go:86] Pod default/nginx-7485d99856-84xqq is unschedulable
I1109 18:34:55.959206       1 klogx.go:86] Pod default/nginx-7485d99856-5q5sw is unschedulable
I1109 18:34:55.959214       1 klogx.go:86] Pod default/nginx-7485d99856-7jlrc is unschedulable
I1109 18:34:55.959223       1 klogx.go:86] Pod default/nginx-7485d99856-cbwq2 is unschedulable
I1109 18:34:55.959231       1 klogx.go:86] Pod default/nginx-7485d99856-b8v2f is unschedulable
I1109 18:34:55.959238       1 klogx.go:86] Pod default/nginx-7485d99856-x2vdf is unschedulable
I1109 18:34:55.959245       1 klogx.go:86] Pod default/nginx-7485d99856-hhzbh is unschedulable
I1109 18:34:55.959253       1 klogx.go:86] Pod default/nginx-7485d99856-bqlz6 is unschedulable
I1109 18:34:55.959261       1 klogx.go:86] Pod default/nginx-7485d99856-p8clz is unschedulable
I1109 18:34:55.959269       1 klogx.go:86] Pod default/nginx-7485d99856-x9m6s is unschedulable
I1109 18:34:55.959276       1 klogx.go:86] Pod default/nginx-7485d99856-955ph is unschedulable
I1109 18:34:55.959284       1 klogx.go:86] Pod default/nginx-7485d99856-w2hdh is unschedulable
I1109 18:34:55.971382       1 klogx.go:86] 39946 other pods are also unschedulable
I1109 18:34:55.971659       1 equivalence_groups.go:40] At the start of buildPodEquivalence Group
I1109 19:26:27.564098       1 equivalence_groups.go:50] Done computing equivalence groups with length: %!s(int=39891)
I1109 19:26:27.564156       1 metrics.go:281] Function buildPodEquivalenceGroups took 51m31.592478269s to complete
```
`buildPodEquivalenceGroup` took 51m31.592478269s in this run.

The `filterOutSchedulable` duration isn't the focus here but I'd like to tackle this next.

- **Second run with the changes in this PR**

```
I1109 20:13:56.833444       1 metrics.go:281] Function filterOutSchedulable took 44m53.991263278s to complete
I1109 20:13:56.833541       1 filter_out_schedulable.go:81] No schedulable pods
I1109 20:13:56.856611       1 klogx.go:86] Pod default/nginx-7485d99856-9cwbx is unschedulable
I1109 20:13:56.856650       1 klogx.go:86] Pod default/nginx-7485d99856-j8lzg is unschedulable
I1109 20:13:56.856656       1 klogx.go:86] Pod default/nginx-7485d99856-wf2qt is unschedulable
I1109 20:13:56.856662       1 klogx.go:86] Pod default/nginx-7485d99856-xlw6c is unschedulable
I1109 20:13:56.856666       1 klogx.go:86] Pod default/nginx-7485d99856-rqsfb is unschedulable
I1109 20:13:56.856672       1 klogx.go:86] Pod default/nginx-7485d99856-pbg86 is unschedulable
I1109 20:13:56.856677       1 klogx.go:86] Pod default/nginx-7485d99856-nmj74 is unschedulable
I1109 20:13:56.856681       1 klogx.go:86] Pod default/nginx-7485d99856-rnjxl is unschedulable
I1109 20:13:56.856686       1 klogx.go:86] Pod default/nginx-7485d99856-z5qcx is unschedulable
I1109 20:13:56.856691       1 klogx.go:86] Pod default/nginx-7485d99856-8qmfq is unschedulable
I1109 20:13:56.856696       1 klogx.go:86] Pod default/nginx-7485d99856-t8nmt is unschedulable
I1109 20:13:56.856700       1 klogx.go:86] Pod default/nginx-7485d99856-ttdrf is unschedulable
I1109 20:13:56.856705       1 klogx.go:86] Pod default/nginx-7485d99856-mf29b is unschedulable
I1109 20:13:56.856710       1 klogx.go:86] Pod default/nginx-7485d99856-rmw9z is unschedulable
I1109 20:13:56.856715       1 klogx.go:86] Pod default/nginx-7485d99856-5r44s is unschedulable
I1109 20:13:56.856720       1 klogx.go:86] Pod default/nginx-7485d99856-f5jnl is unschedulable
I1109 20:13:56.856725       1 klogx.go:86] Pod default/nginx-7485d99856-twc9f is unschedulable
I1109 20:13:56.856729       1 klogx.go:86] Pod default/nginx-7485d99856-w2st5 is unschedulable
I1109 20:13:56.856734       1 klogx.go:86] Pod default/nginx-7485d99856-f7vp6 is unschedulable
I1109 20:13:56.856739       1 klogx.go:86] Pod default/nginx-7485d99856-6ck5x is unschedulable
I1109 20:13:56.865525       1 klogx.go:86] 39946 other pods are also unschedulable
I1109 20:13:56.865784       1 equivalence_groups.go:40] At the start of buildPodEquivalence Group
I1109 20:13:58.577666       1 equivalence_groups.go:50] Done computing equivalence groups with length: %!s(int=1)
I1109 20:13:58.577713       1 metrics.go:281] Function buildPodEquivalenceGroups took 1.71191175s to complete
```

You can see that the `filterOutSchedulable` has decreased significantly for that run. I think without that change the inner loop is exponential as you keep adding more groups as you're looping through the 40k pods.

Reply:
@MaciekPytel gentle ping on the above.

Reply:
@MaciekPytel @gjtempleton @mwielgus ping?

Reply:
That's very interesting results. I'm not sure if the particular scenario you tested is what we want to optimize for though. 

Historically it haven't been. Our idea for CA scalability was always to keep up with the scheduler. Since scheduler has limited mutating QPS to api-server it can only mark so many pods as unschedulable during the time it takes CA to run a single loop. If we keep loops fast enough we never have to deal with thousands of pending pods in one go. We tested (though only on GCP, I don't have access to that much compute elsewhere) adding more than 40k pods in one go, but with CA already running. In that scenario the number of pending pods is rate-limited by scheduler marking pods as unschedulable.
The question is how much do we need CA to be able to handle a huge number of already pending pods vs preventing the pod buildup in the first place. I think both require different set of optimizations.

FilterOutSchedulable is a great example here. In principle you could add something like equivalenceGroups to cache negative results (pods can't schedule on existing nodes) and it would probably massively help in that case. We never did it, because we never needed to. If CA keeps up with scheduler there should never be a buildup of enough pods for it to matter. The caching of positive results between loops on the other hand is critical. Even if we keep up and every scale-up only ever needs to handle a few hundred pending pods, filterOutSchedulable has to contend with all the pods that are waiting for VMs to boot up. VM boot is long enough to allow tens of thousands of pods to build up, all waiting for upcoming nodes. But thanks to positive caching, in each loop there are only a few hundred pods that need to be scheduled, the rest can be assigned based on caching from previous loops. 

Your optimization is clearly a welcome improvement. I'm not as sure if it makes quite that much difference in a scenario where CA is running from the very beginning and can keep up with scheduler. I would expect the biggest problem to be in estimator (I'd love to see how much time was spent in it in your test?). It's running filters O(#pending_pods * #nodes_delta) times and since #node_delta is a function of #pending_pods, it's really O(#pending_pods ^ 2). With so many pods I can't imagine it being fast.
You have a good point though, with that many different equivalenceGroups in a single controller we run a whole lot of DeepEquals (O(#pending_pods ^ 2) I think?). It may very well be that those DeepEquals are what makes up all this time in buildEquivalenceGroups. So the total time spent in estimator may turn out to not be that much higher (but likely harder to fix I think) in this specific scenario.

Reply:
> Historically it haven't been. Our idea for CA scalability was always to keep up with the scheduler. Since scheduler has limited mutating QPS to api-server it can only mark so many pods as un-schedulable during the time it takes CA to run a single loop. If we keep loops fast enough we never have to deal with thousands of pending pods in one

Yeah that makes sense. The reason I came up across is that I was running some automated large scenario testing with the autoscaler health-check enabled. Obviously since the main loop gets blocked, it failed the liveness probe and eventually enters a never-ending loop of restarts which build up the pending pod pressure. I get the scheduler QPS point but those parameters are configurable but it's not uncommon to see people increase those and the API server mutating QPS in large cluster scenarios - at the cost of more cpu and memory sometimes. I am also thinking of a scenario where, for whatever reason, the CA process crashes and ends up having to catch up with more work.

> FilterOutSchedulable is a great example here. In principle you could add something like equivalenceGroups to cache negative results (pods can't schedule on existing nodes) and it would probably massively help in that case. We never did it, because we never needed to.

That was another thing I was going to look into at after with equivalence groups there. 

On that note, I did see `FilterOutSchedulable` complete significantly faster with CA 1.17 compared to the latest master. I probably won't get to do another run till the new year but it was something I wanted to dig into. 

>  If CA keeps up with scheduler there should never be a buildup of enough pods for it to matter. The caching of positive results between loops on the other hand is critical. 

Yup - I did the see the effect of the positives pretty clearly in the later runs and I agree its definitely the more critical part.

>  I'm not as sure if it makes quite that much difference in a scenario where CA is running from the very beginning and can keep up with scheduler

It won't make too much of a difference in the happy path for sure.

> I would expect the biggest problem to be in estimator (I'd love to see how much time was spent in it in your test?)
 It's running filters O(#pending_pods * #nodes_delta) times and since #node_delta is a function of #pending_pods, it's really O(#pending_pods ^ 2). With so many pods I can't imagine it being fast.

The estimator is one thing that surprised me. Does that mean that we are simply just seeing the cost of the `DeepEquals` calls since estimator runs seem much faster than anticipated? 

```
I1104 17:38:16.087658       1 scale_up.go:410] Calling estimate for nodeGroup:nodepool1-32591543 with number of pods: 39966
I1104 17:39:40.231778       1 metrics.go:293] Function Estimate took 1m24.144072761s to complete
```

It took around 1.5 minute per nodegroup and later ones were faster as the number of pods to binpack kept dropping.

```
I1104 17:45:48.171285       1 scale_up.go:410] Calling estimate for nodeGroup: nodepool1-32591543 with number of pods: 30975
I1104 17:46:35.787187       1 metrics.go:293] Function Estimate took 47.615854365s to complete
```

Reply:
Great convo and great work @marwanad

Reply:
> On that note, I did see FilterOutSchedulable complete significantly faster with CA 1.17 compared to the latest master. I probably won't get to do another run till the new year but it was something I wanted to dig into.

That's not unexpected. In 1.18 the scheduler moved onto 'scheduler framework' and to remain compatible we had to completely rewrite scheduler integration in CA and a lot of associated code (including the entire filterOutSchedulable logic). The new version is arguably more correct, but we've lost some of our old optimizations (ex. the ability to completely disable InterPodAffinity logic if there are no pods in the cluster using it). 

> It took around 1.5 minute per nodegroup

That is very interesting. After some digging through the code, I reached the following conclusions:
 * It definitely makes me suspect the buildPodEquivalenceGroup cost is not due to filters. DeepEqual seems like the likely culprit, but I'm not sure if there is any way to be completely sure other than instrumenting / profiling it and running the test.
 * Turns out I was completely wrong about filterOutSchedulable - we do negative caching, which seems very similar to the one used in podEquivalenceGroups: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/filter_out_schedulable.go#L150. And it also runs a ton of DeepEqual in your scenario (we essentially has a hash map, where every pod has a hash conflict and it degenerates into a list).
   * I bet changing https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/utils/pod_schedulable.go#L53 to PodSpecSemanticallyEqual is going to fix the problem with filterOutSchedulable in your test. Would you mind testing and, if it works, adding it to this PR? Alternatively, we could merge this one and do it in a follow-up.
   * I think the fact that we have 2 implementations of essentially the same caching logic is not great. This PR is a great example of how easily an optimization can get applied to one implementation, but not the other. I think the ideal state would be to try and merge the two implementations (and maybe consider using them in estimator too? I see no reason why negative caching wouldn't be applicable there). Obviously this is a much bigger refactor and not in scope of this PR.

Reply:
> scheduler framework
yup - I remember the check to skip the pod affinity predicate since it was costly.

`PodSemanticallyEqual` still uses `DeepEquals` right? I think the culprit the combination of that #pending_pods^2 factor and `DeepEqual` call all bundled together.

With the same setup, I did one run with the main branch and didn't wait for it to complete for the interest of time. At 30 mins, it was still running.

After changing the call `PodSemanticallyEqual` (and thus reducing the lists):
```
I1220 21:18:01.664867       1 filter_out_schedulable.go:65] Filtering out schedulables
I1220 21:18:05.237518       1 metrics.go:322] Function filterOutSchedulable took 3.572606755s to complete
```

I updated the PR with that change too.

Good catch there! That can probably give us some seconds back in medium-scale scenarios too for those K8s versions.

I let it run for a bit to get estimator data again as well:
```
I1220 21:38:34.468866       1 filter_out_schedulable.go:65] Filtering out schedulables
I1220 21:38:36.696208       1 metrics.go:322] Function filterOutSchedulable took 2.227283306s to complete
I1220 21:38:36.720877       1 klogx.go:86] Pod default/nginx-867649cf6c-vq5vg is unschedulable
I1220 21:38:36.720938       1 klogx.go:86] Pod default/nginx-867649cf6c-wcns6 is unschedulable
I1220 21:38:36.720945       1 klogx.go:86] Pod default/nginx-867649cf6c-zsxlm is unschedulable
I1220 21:38:36.720950       1 klogx.go:86] Pod default/nginx-867649cf6c-4jvwm is unschedulable
I1220 21:38:36.720955       1 klogx.go:86] Pod default/nginx-867649cf6c-g6qqc is unschedulable
I1220 21:38:36.720960       1 klogx.go:86] Pod default/nginx-867649cf6c-5khff is unschedulable
I1220 21:38:36.720965       1 klogx.go:86] Pod default/nginx-867649cf6c-xr8x2 is unschedulable
I1220 21:38:36.720970       1 klogx.go:86] Pod default/nginx-867649cf6c-r6bqh is unschedulable
I1220 21:38:36.720975       1 klogx.go:86] Pod default/nginx-867649cf6c-ntsm5 is unschedulable
I1220 21:38:36.720980       1 klogx.go:86] Pod default/nginx-867649cf6c-f6r67 is unschedulable
I1220 21:38:36.720985       1 klogx.go:86] Pod default/nginx-867649cf6c-g58kg is unschedulable
I1220 21:38:36.720990       1 klogx.go:86] Pod default/nginx-867649cf6c-nmht2 is unschedulable
I1220 21:38:36.720996       1 klogx.go:86] Pod default/nginx-867649cf6c-tsbpl is unschedulable
I1220 21:38:36.721001       1 klogx.go:86] Pod default/nginx-867649cf6c-7v9zc is unschedulable
I1220 21:38:36.721006       1 klogx.go:86] Pod default/nginx-867649cf6c-g4bhj is unschedulable
I1220 21:38:36.721017       1 klogx.go:86] Pod default/nginx-867649cf6c-t5pgb is unschedulable
I1220 21:38:36.721022       1 klogx.go:86] Pod default/nginx-867649cf6c-2ftsx is unschedulable
I1220 21:38:36.721027       1 klogx.go:86] Pod default/nginx-867649cf6c-g8wsf is unschedulable
I1220 21:38:36.721032       1 klogx.go:86] Pod default/nginx-867649cf6c-lkblx is unschedulable
I1220 21:38:36.721037       1 klogx.go:86] Pod default/nginx-867649cf6c-5lv2b is unschedulable
I1220 21:38:36.740616       1 klogx.go:86] 39795 other pods are also unschedulable
I1220 21:38:40.044023       1 equivalence_groups.go:50] Done computing equivalence groups with length: 1
I1220 21:38:40.044092       1 metrics.go:322] Function buildPodEquivalenceGroups took 3.303054365s to complete
I1220 21:38:43.888896       1 metrics.go:322] Function Estimate took 3.844386108s to com
I1220 21:38:40.044485       1 binpacking_estimator.go:65] At the start of Estimate
I1220 21:38:40.044485       1 binpacking_estimator.go:118] Function Estimate took 7m53.871246475s to complete
```

> I think the ideal state would be to try and merge the two implementations (and maybe consider using them in estimator too? I see no reason why negative caching wouldn't be applicable there).

Yeah there's definitely room for logic sharing there. I'll try and revisit this point again in the new year.

Reply:
> scheduler framework
yup - I remember the check to skip the pod affinity predicate since it was costly.

`PodSemanticallyEqual` still uses `DeepEquals` right? I think the culprit the combination of that #pending_pods^2 factor and `DeepEqual` call all bundled together.

With the same setup, I did one run with the main branch and didn't wait for it to complete for the interest of time. At 30 mins, it was still running.

After changing the call to `PodSemanticallyEqual` in podschedulable info (and thus reducing the lists):

Complete run with both changes:
```
I1220 21:38:34.468866       1 filter_out_schedulable.go:65] Filtering out schedulables
I1220 21:38:36.696208       1 metrics.go:322] Function filterOutSchedulable took 2.227283306s to complete
I1220 21:38:36.720877       1 klogx.go:86] Pod default/nginx-867649cf6c-vq5vg is unschedulable
I1220 21:38:36.720938       1 klogx.go:86] Pod default/nginx-867649cf6c-wcns6 is unschedulable
I1220 21:38:36.720945       1 klogx.go:86] Pod default/nginx-867649cf6c-zsxlm is unschedulable
I1220 21:38:36.720950       1 klogx.go:86] Pod default/nginx-867649cf6c-4jvwm is unschedulable
I1220 21:38:36.720955       1 klogx.go:86] Pod default/nginx-867649cf6c-g6qqc is unschedulable
I1220 21:38:36.720960       1 klogx.go:86] Pod default/nginx-867649cf6c-5khff is unschedulable
I1220 21:38:36.720965       1 klogx.go:86] Pod default/nginx-867649cf6c-xr8x2 is unschedulable
I1220 21:38:36.720970       1 klogx.go:86] Pod default/nginx-867649cf6c-r6bqh is unschedulable
I1220 21:38:36.720975       1 klogx.go:86] Pod default/nginx-867649cf6c-ntsm5 is unschedulable
I1220 21:38:36.720980       1 klogx.go:86] Pod default/nginx-867649cf6c-f6r67 is unschedulable
I1220 21:38:36.720985       1 klogx.go:86] Pod default/nginx-867649cf6c-g58kg is unschedulable
I1220 21:38:36.720990       1 klogx.go:86] Pod default/nginx-867649cf6c-nmht2 is unschedulable
I1220 21:38:36.720996       1 klogx.go:86] Pod default/nginx-867649cf6c-tsbpl is unschedulable
I1220 21:38:36.721001       1 klogx.go:86] Pod default/nginx-867649cf6c-7v9zc is unschedulable
I1220 21:38:36.721006       1 klogx.go:86] Pod default/nginx-867649cf6c-g4bhj is unschedulable
I1220 21:38:36.721017       1 klogx.go:86] Pod default/nginx-867649cf6c-t5pgb is unschedulable
I1220 21:38:36.721022       1 klogx.go:86] Pod default/nginx-867649cf6c-2ftsx is unschedulable
I1220 21:38:36.721027       1 klogx.go:86] Pod default/nginx-867649cf6c-g8wsf is unschedulable
I1220 21:38:36.721032       1 klogx.go:86] Pod default/nginx-867649cf6c-lkblx is unschedulable
I1220 21:38:36.721037       1 klogx.go:86] Pod default/nginx-867649cf6c-5lv2b is unschedulable
I1220 21:38:36.740616       1 klogx.go:86] 39795 other pods are also unschedulable
I1220 21:38:40.044023       1 equivalence_groups.go:50] Done computing equivalence groups with length: 1
I1220 21:38:40.044092       1 metrics.go:322] Function buildPodEquivalenceGroups took 3.303054365s to complete
I1220 21:38:40.044485       1 binpacking_estimator.go:65] At the start of Estimate
I1220 21:38:43.888896       1 metrics.go:322] Function Estimate took 7m53.871246475s to complete
```

I updated the PR with that change too.

Good catch there! That can probably give us some seconds back in medium-scale scenarios too for those K8s versions.

> I think the ideal state would be to try and merge the two implementations (and maybe consider using them in estimator too? I see no reason why negative caching wouldn't be applicable there).

Yeah there's definitely room for logic sharing there. I'll try and revisit this point again in the new year.

Reply:
> PodSemanticallyEqual still uses DeepEquals right? I think the culprit the combination of that #pending_pods^2 factor and DeepEqual call all bundled together.

Yeah, that's what I meant. If all those pods end up as a separate entry in podSchedulableMap, each pod is compared with each previously processed pod. With PodSemanticallyEqual there is just one comparison needed per pod.

I wonder if we could come up with some sort of fallback. The caching in podSchedulableMap and equivalenceGroups is meant as an optimization, but as this case demonstrated it can become prohibitively expensive instead in cases where there are thousands of pods with different scheduling requirements owned by the same controller. Regardless of why those pods are different, at some point the caching becomes too expensive to be worth it. Maybe we could come up with a cutoff point, where we stop caching once the list of 'different' pods from the same controller grows too long? An example implementation could be something like adding `if len(podMap[uid]) < maxLength` in https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/utils/pod_schedulable.go#L80.

Reply:
hmm interesting idea - so beyond that maxLength - we just consider the remaining pods different?

Reply:
Yeah, basically once we reach a certain list length we don't append anymore. We risk missing out on some caching opportunity, but once the list gets long enough the cost of comparisons exceeds the cost of running Filters and so the caching hurts us more than it helps.

In practice, my guess is that either all pods in a given controller are identical (or there are 2 variants in case of some ongoing update) or there is some issue which makes them all different (like the one we have here). I can't really think of a case where a single controller has a 100+ "classes" of pods, with multiple pods in each class. So my guess is that if the list gets very long, it probably means we don't get anything out of caching anyway.